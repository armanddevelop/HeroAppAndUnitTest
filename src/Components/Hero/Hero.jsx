import { useMemo } from "react";
import { useParams, Navigate, useNavigate } from "react-router-dom";
import { getHeroById } from "../../selectors/getHeroById";
const herosImages = require.context("../../assets/heros", true);

export const Hero = () => {
  const { heroId } = useParams();
  console.log("esto vael heroId ", heroId);
  const navigate = useNavigate();
  const heroData = useMemo(() => getHeroById(heroId), [heroId]);

  if (Object.entries(heroData).length === 0) {
    return <Navigate to="/" />;
  }
  const imgPath = herosImages(`./${heroData.id}.jpg`);
  const handleReturn = () => {
    const path = heroId.split("-");
    navigate(`/${path[0]}`, { replace: true });
  };
  return (
    <div className="row-mt-5">
      <h1>{heroData.characters}</h1>
      <div className="col-4">
        <img
          src={imgPath}
          alt={heroData.superhero}
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>
      <div className="col-8 animate__animated animate__fadeIn">
        <h3>{heroData.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Aleter ego:</b> {heroData.alter_ego}
          </li>
          <li className="list-group-item">
            <b>Bio:</b>
            <p>{heroData.information}</p>
          </li>
          <li className="list-group-item">
            <b>Publisher:</b>
            <p>{heroData.publisher}</p>
          </li>
          <li className="list-group-item">
            <b>First Appearence:</b>
            <p>{heroData.first_appearance}</p>
          </li>
        </ul>
        <h5 className="mt-3">Characters</h5>
        <p>{heroData.characters}</p>
        <button className="btn btn-outline-info" onClick={handleReturn}>
          Return
        </button>
      </div>
    </div>
  );
};
