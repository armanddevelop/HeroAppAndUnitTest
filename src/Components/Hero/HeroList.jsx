import PropTypes from "prop-types";
import { useMemo } from "react";
import { getHeroByPublisher } from "../../selectors/getHeroByPublisher";
import { HeroCard } from "./HeroCard";

export const HeroList = ({ publisher }) => {
  const heroList = useMemo(() => getHeroByPublisher(publisher), [publisher]);
  return (
    <div className="row rows-cols-1 row-cols-md-3 g-3 animate__animated animate__fadeIn">
      {heroList.map((hero) => (
        <HeroCard key={hero.id} {...hero} />
      ))}
    </div>
  );
};

HeroList.propTypes = {
  publisher: PropTypes.string.isRequired,
};
