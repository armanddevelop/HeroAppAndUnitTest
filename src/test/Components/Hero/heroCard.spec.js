import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { HeroCard } from "../../../Components/Hero/HeroCard";

describe("Test to HeroCard", () => {
  const hero = {
    id: "dc-batman",
    superhero: "batman",
    publisher: "DC Comics",
    alter_ego: "paquito",
    first_appearance: "1925",
    characters: "bruce wayne",
  };
  const container = shallow(<HeroCard {...hero} />);
  it("SnapShot to HeroCard component", () => {
    expect(toJson(container)).toMatchSnapshot();
  });
  it("alter_ego label", () => {
    const hero = {
      id: "dc-batman",
      superhero: "batman",
      publisher: "DC Comics",
      alter_ego: "bruce wayne",
      first_appearance: "1925",
      characters: "bruce wayne",
    };
    const container = shallow(<HeroCard {...hero} />);
    const textMuted = container.find(".text-muted").length;
    expect(textMuted).toBe(1);
  });
  it("alter ego label", () => {
    const hero = {
      id: "dc-batman",
      superhero: "batman",
      publisher: "DC Comics",
      alter_ego: "bruce wayne",
      first_appearance: "1925",
      characters: "bruce wayne",
    };
    const container = shallow(<HeroCard {...hero} />);
    const textMuted = container.find(".text-muted").length;
    expect(textMuted).toBe(1);
  });
  it("characters label", () => {
    const textMuted = container.find(".text-muted").length;
    expect(textMuted).toBe(2);
  });
  it("link label", () => {
    const link = container.find("Link").exists();
    expect(link).toBeTruthy();
  });
});
