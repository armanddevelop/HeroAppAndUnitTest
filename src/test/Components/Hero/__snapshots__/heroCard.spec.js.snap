// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to HeroCard SnapShot to HeroCard component 1`] = `
<div
  className="col animate__animated animate__fadeIn"
>
  <div
    className="card"
  >
    <div
      className="row no-gutters"
    >
      <div
        className="col-4"
      >
        <img
          alt="batman"
          className="card-img"
          src="/assets/heroes/dc-batman.jpg"
        />
      </div>
      <div
        className="col-8"
      >
        <div
          className="card-body"
        >
          <h5
            className="card-title"
          >
            batman
          </h5>
          <p
            className="card-text"
          >
            paquito
          </p>
          <p
            className="text-muted"
          >
            bruce wayne
          </p>
          <p
            className="card-text"
          >
            <small
              className="text-muted"
            >
              1925
            </small>
          </p>
          <Link
            to="/hero/dc-batman"
          >
            Mas Info
          </Link>
        </div>
      </div>
    </div>
  </div>
</div>
`;
